File: src/band_ack/band_ack.c
#include <stdio.h>



/// Get the ticks counter on the devices with the provided address
///
/// address     Address of the target device, configured by dip switches
/// device      Path to the serial device to use
/// command     Specifies the action we perform
///             0 ask for the ticks
///             1 ping
/// ticks       Pointer to a 16 bit var which will contain the current tick counter
///
/// return      0 No error
///             1 No answer from device
///             2 Invalid answer
///             3 ...
unsigned char call_band_ack(unsigned char address, char* device, unsigned char command, unsigned short *ticks) {
  printf("get_ticks address: %i, command: %i, with device %s\n", address, command, device);
  *ticks = 2;
  return 0;
}

File: src/band_ack/band_ack.rs
use std::ffi::CString;
use std::os::raw::c_char;
use band_ack::error::TickError;



extern {
    fn call_band_ack(address: u8, device: *const c_char, command: u8, ticks: &u16) -> u8;
}



/// Get the ticks counter on the devices with the provided address
///
/// address     Address of the target device, configured by dip switches
/// device      Path to the serial device to use
/// return      Error or ticks
pub fn get_ticks(address: u8, device: String) -> Result<u16, TickError> {
    let ticks = 0_u16;
    let d = CString::new(device).unwrap();
    let error = unsafe { call_band_ack(address, d.as_ptr(), 1_u8, &ticks) };
    if let Some(error) = TickError::from_u8(error) {
        return Err(error)
    }
    return Ok(ticks)
}
